Diagnostic/diagnostic.py

```diff
@@ -828,6 +848,21 @@ def start_mdsd():
 
                 # mdsd is now up for at least 30 seconds.
 
+                # Mitigate if memory leak is suspected.
+                mdsd_memory_leak_suspected, mdsd_memory_usage_in_MB = check_suspected_memory_leak(mdsd.pid)
+                if mdsd_memory_leak_suspected:
+                    memory_leak_msg = "Suspected mdsd memory leak (memory usage: {0}MB). Recycling mdsd to self-mitigate.".format(mdsd_memory_usage_in_MB)
+                    hutil.log(memory_leak_msg)
+                    # Add a telemetry as well. This might ding our enable success KPI,
+                    # but there doesn't seem to be any other way.
+                    waagent.AddExtensionEvent(name=hutil.get_name(),
+                                              op=ExtensionOperationType,
+                                              isSuccess=False,
+                                              version=hutil.get_extension_version(),
+                                              message=memory_leak_msg)
```

Comment:
I disagree with setting the status to "failed". As a result of the mitigation, the service continues to run. I agree with the need for telemetry, but why can't you AddExtensionEvent with success==true? What is the total space of ExtensionOperationType - if we can add a new "extension operation type" which corresponds to "auto-restart", then we can log success for that new operation and easily (I hope) do analysis of the events reported in this way.